### PR TITLE
Memory constraints and constraint handling refactor

### PR DESCRIPTION
**Pull request by grjte**
This PR implements the memory constraints (excluding permutation checks) and the top level aux table constraints, introduces a possible new structure for handling constraints, and refactors the range checker constraints to the new structure.

While I maintained many aspects of the approach from the previous version, I wanted to explore a few different ideas to address the following:

1. Think about different ways to approach flags for turning on/off specific constraints or entire processors.
2. Make it easy to test systematically
3. Have a systematic structure that will let us derive or generate aspects of the constraint-handling via macros. (For example I think the column indices, the column accessors in the frame extension trait, and the TransitionConstraint builders could all be reduced this way.)
4. Try co-locating degree declarations and constraint definitions rather than having them completely decoupled, since I thought this might be easier to maintain, test, and understand if we can find a nice way to do it.
5. Reduce the number of assertions at the top level and contain them in modules, since we have more processors and constraints in this version.
6. Avoid specifying constraint evaluation indices manually when setting the result array, since I think this could be error prone, especially as the number of constraints grows.

I tried to layer the different ideas by commit (assuming we may not keep all of them), with the first/second commit being closest to the previous approach.

The processor constraints structs are just to give a single interface, with ColumnBoundary intended to make it easy to see the list of boundary conditions for a processor by column. The frame extension trait lets us avoid duplicating variable computations so we can reuse in different constraints and test easily, as well as construct function pointers for constraint evaluations. The TransitionConstraint + builder let us compose an optional flag with a constraint for evaluation, as well as co-locating the degree so all details of a constraint are defined in one place.  This is also what lets us stop specifying indices manually in the result array and not worry about having the index match the declaration of the constraint degree array or about making mis-indexing errors that slip through the cracks because they don't cause a degree mismatch.

### CODE REVIEW

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you so much! This is great work. I left some minor comments inline, but the meat of my thoughts is below. Also, I didn't review the actual memory constraints in detail as the overall structure is likely to change.

- I really like the idea of extending evaluation frame interface. This makes writing constraints less error-prone, and makes reading them much easier.
- I also like the idea of having separate structs which encapsulate constraints for a given segment. However, I noticed that you have two structs, one for boundary constraints (e.g., `RangeCheckerBoundaries`) and another one for transition constraints (e.g., `RangeCheckerConstraints`). I would have probably gone with a single struct for both - especially since each struct is fairly simple.
- I do like the idea of breaking up constraints into different functions, but I think having one function per constraint might be a bit too much. The main reason is that reading constraints was a bit more difficult for me this time (comparing current vs. past range checker constraints). For something as simple as range checker constraints one function is probably enough. For more complicated ones (e.g., memory) we probably can break them up into several functions.
- I also like the idea of tying constraint degree declaration with the actual constraint and `ProcessorTransitions` structure looks very cool. But I fear that function pointer approach may result in noticeable performance degradation. So, until we can measure the impact of this approach, we should probably stay away from it.

To summarize:
- Let's keep the frame interface extensions
- Let's keep the overall hierarchical structure that you have, but let's have a single struct per logical set of constraints which combines both boundary and transition constraints.
- Let's group transition constraints a bit more (not one constraint be function).
- For now, let's roll back the function pointer-based transition constraint declaration/evaluation.

I understand that the structure will be far away from ideal, and I do think we'll make quite a few refactorings going forward - but I'd rather take a slightly more iterative approach.

**Comment by bobbinth** [comment]
One other thing that is worth mentioning is the whole "single assignment" vs. "multi-assignment" approach. Let's take the auxiliary table for example. Let's say our memory segment has 20 constraints and hasher segment has 20 constraints. We know that these constraints are fully disjointed (e.g., each set of constraints is under a flag, and these flags are mutually exclusive).

If I understood it correctly, with the approach of this PR we would need 40 slots in the `results` slice to record these constraints. However, since all constraints are disjoined, we could get away with only 20 slots. The "cost" of each slot is basically 3 multiplications - so, the approach which uses 40 slots would require 60 more multiplications - which may be noticeable. This situation becomes even more relevant when we have multiple sets of constraints (e.g. 20 constraints for hasher, 20 for memory, 20 for bitwise etc.) which can be "packed" into a single result slot.

File: air/src/lib.rs [hunk 1]
```diff
@@ -93,14 +127,46 @@ impl Air for ProcessorAir {
         result[0] = next[CLK_COL_IDX] - (current[CLK_COL_IDX] + E::ONE);
 
         // --- range checker ----------------------------------------------------------------------
-        range::enforce_constraints::<E>(frame, &mut result[1..]);
+        RangeCheckerConstraints::new()
+            .enforce_constraints(frame, &mut result[self.range_checker_start()..]);
+
+        // --- auxiliary table of co-processors (hasher, bitwise, memory) -------------------------
+        AuxTableConstraints::new()
+            .enforce_constraints(frame, &mut result[self.aux_table_start()..]);
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/lib.rs:135]: It wasn't possible to call enforce_constraints on an instance variable from evaluate_transition because of the generic type. Also, the previous version was instantiating the VM transition every time, so I thought this might be the preferred approach vs having an instance variable.

Once we do away with the function-based generic transition constraint approach this should be doable, and it's my preferred approach as well, given that you don't have objections to adding these structs as instance variables.

File: air/src/aux_table/mod.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    fn enforce_constraints(&self, frame: &EvaluationFrame<E>, result: &mut [E]) {
+        // --- auxiliary table management ---------------------------------------------------------
+        for (index, constraint) in self.transitions().iter().enumerate() {
+            result[index] = constraint.evaluate(frame);
+        }
+
+        // --- memory -----------------------------------------------------------------------------
+        self.memory
+            .enforce_constraints(frame, &mut result[self.memory_start()..]);
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/aux_table/mod.rs:104]: I took this approach in a previous commit as well. However, I removed that when I switched to the approach with the processor constraints, because most of the structs using this trait won't need it, so passing it here didn't match the signature. If you don't like the current handling accessing the AuxTable frame extension then I could pass the flag when instantiating the memory constraints instead. What do you think?

**Comment by grjte** [comment]
Great, thanks!

>     * I do like the idea of breaking up constraints into different functions, but I think having one function per constraint might be a bit too much. The main reason is that reading constraints was a bit more difficult for me this time (comparing current vs. past range checker constraints). For something as simple as range checker constraints one function is probably enough. For more complicated ones (e.g., memory) we probably can break them up into several functions.

Yeah, I agree that it became harder to verify individual constraints with this latest approach. I think part of this was because of the extra layers of direction to check the variables and flags. In a [previous commit](https://github.com/maticnetwork/miden/pull/173/commits/390921d976d54bec5f575606747d1474204862c6)  I had constraints grouped by purpose I.e. in the memory constraints I enforced d_inv, delta, and values. I think this was more readable as well.

The reasons I tried changing it were to: 
1. Associate the degree with the constraint directly.
2. Allow testing each individual constraint easily.
3. Avoid manually indexing into the result array, potentially making indexing errors.

If this 1-function-per-constraint approach didn't have this readability problem, would you prefer it for these reasons it would you still prefer the other approach? Because removing the function pointers should make it possible/easier to reduce this issue.

>    I fear that function pointer approach may result in noticeable performance degradation. So, until we can measure the impact of this approach, we should probably stay away from it.

Yep, makes sense, especially with the effects on compiler inlining.

**Comment by grjte** [comment]
> One other thing that is worth mentioning is the whole "single assignment" vs. "multi-assignment" approach. 

This makes a lot of sense, and is not something I had considered as a design decision. If we were to stop consider taking a single-function-per-constraint approach for the reasons mentioned above, I think we could group constraints by exclusive flags so that we could combine them while still setting clearly what's going on. Although this aggregation is not done in this version, there's no reason it couldn't have been.

Whether we take the single function or multi-function approach, it would be good to make it very clear what's going on with these multi-assignment combinations, because it's obviously worth doing for performance reasons, but can be hard to follow

**Review by bobbinth — CHANGES_REQUESTED** [review]
Sorry for another partial review - but I won't be able to finish the whole thing today. I left some inline comments. Most of them are related to reducing the amount of boilerplate code.

File: air/src/lib.rs [hunk 3]
```diff
@@ -24,24 +32,53 @@ pub struct ProcessorAir {
     context: AirContext<Felt>,
     stack_inputs: Vec<Felt>,
     stack_outputs: Vec<Felt>,
+    constraint_ranges: TransitionConstraintRange,
+    range_constraints: RangeCheckerConstraints,
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/lib.rs:36]: Yeah, this bothered me too. Maybe just switching it to "range_checker" will be best

File: air/src/lib.rs [hunk 4]
```diff
@@ -24,24 +32,53 @@ pub struct ProcessorAir {
...
+        // --- auxiliary table of co-processors (hasher, bitwise, memory) -------------------------
+        let aux_table_constraints = AuxTableConstraints::new();
+        degrees.append(&mut aux_table_constraints.get_transition_constraint_degrees());
+
+        // Define the transition constraint ranges.
+        let constraint_ranges = TransitionConstraintRange::new(
+            1,
+            range_constraints.get_transition_constraint_count(),
+            aux_table_constraints.get_transition_constraint_count(),
+        );
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/lib.rs:73]: Yeah, I thought about this as well, but was aiming for something well-structured. Revisiting this once all the constraints are done would make sense.

File: air/src/range.rs [hunk 5]
```diff
@@ -1,75 +1,183 @@
...
-    result[3] = (E::ONE - next[T_COL_IDX]) * (row_diff) * (row_diff - E::ONE);
+    constraint_count
+}
 
-    // --- Constrain the transition from 8-bit to 16-bit section of the table. --------------------
-    let flip = (E::ONE - current[T_COL_IDX]) * next[T_COL_IDX];
+/// Constrain the transition from 8-bit to 16-bit section of the table.
+fn enforce_16bit<E: FieldElement>(frame: &EvaluationFrame<E>, result: &mut [E]) -> usize {
+    let constraint_count = 3;
+    let flip = binary_not(frame.t()) * frame.t_next();
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/range.rs:111]: Yes, I think there's a good argument for having all intermediate variables and flags defined in EvaluationFrameExt. It's not necessary for the ones only used in one place, so I was undecided if it was overkill or not. However, it's useful to have a single place where everything is defined

**Comment by grjte** [comment]
> Most of them are related to reducing the amount of boilerplate code.

Just to add context here - my original thinking had been that once we had a well-defined structure that was consistent across all modules we could just use a proc macro to derive from a very minimal declaration of the core info required by that module, after which we wouldn't actually see the boilerplate. (This doesn't mean that there aren't things that can be reduced more now)

File: air/src/range.rs [hunk 6]
```diff
@@ -1,75 +1,183 @@
...
+            first_step,
+            last_step,
+            transitions,
+        }
+    }
 }
 
-// BOUNDARY CONSTRAINTS
-// ================================================================================================
+impl ProcessorConstraints for RangeCheckerConstraints {
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/range.rs:57]: My thinking had been that we could have the transition & boundary handling methods defined in the ProcessorConstraints, add a derive macro on each of the individual module structs, and then only handle initialization and `enforce_constraints` in each of the modules (except for the auxiliary table which needs to override more things for handling coprocessors) while having a consistent interface in the Air impl.

We definitely don't have to take this approach. It's just what I had in mind when I set it up this way, so this PR is basically an intermediate step, because it's not worth doing the macro if you don't like this general structure. Let me know what you think - you have a much better sense of what's coming next with the rest of the constraint handling

[the remark this answers is not in the document]
Comment (grjte) [reply air/src/range.rs:57]: In fact, I had been thinking that possibly even the initialization could be derived, if there's a good way to declare our boundary constraints. In that case, we would just have the declaration of our `NUM_CONSTRAINTS`, `CONSTRAINT_DEGREES`, boundaries (somehow), and then the `enforce` functions and frame extension

File: air/src/range.rs [hunk 7]
```diff
@@ -1,75 +1,183 @@
...
-pub fn get_assertions_first_step(result: &mut Vec<Assertion<Felt>>) {
-    result.push(Assertion::single(V_COL_IDX, 0, Felt::ZERO));
-}
+    fn first_step(&self) -> &[ColumnBoundary] {
+        &self.first_step
+    }
+
+    fn last_step(&self) -> &[ColumnBoundary] {
+        &self.last_step
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/range.rs:66]: This is related to the [discussion above ](https://github.com/maticnetwork/miden/pull/173#discussion_r851112285) about using a macro to reduce all modules to simple declarations of transition/boundary constraints, frame extension trait (which can also be partially derived if we want), and enforce methods.

If we don't do that, then I think this makes sense.

File: air/src/range.rs [hunk 8]
```diff
@@ -1,75 +1,183 @@
...
+
+    // --- TRANSITION CONSTRAINTS -----------------------------------------------------------------
 
-pub fn get_assertions_last_step(result: &mut Vec<Assertion<Felt>>, step: usize) {
-    result.push(Assertion::single(V_COL_IDX, step, Felt::new(65535)));
+    fn transitions(&self) -> &TransitionConstraints {
+        &self.transitions
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/range.rs:72]: This is related the [above discussion](https://github.com/maticnetwork/miden/pull/173#discussion_r851112285)

File: air/src/lib.rs [hunk 9]
```diff
@@ -93,14 +131,49 @@ impl Air for ProcessorAir {
         result[0] = next[CLK_COL_IDX] - (current[CLK_COL_IDX] + E::ONE);
 
         // --- range checker ----------------------------------------------------------------------
-        range::enforce_constraints::<E>(frame, &mut result[1..]);
+        self.range_constraints
+            .enforce_constraints::<E>(frame, &mut result[self.range_start()..]);
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/lib.rs:135]: We can't do that, because std::ops::Range doesn't implement Copy. I could just clone it though

**Comment by grjte** [comment]
As a very rough sketch, the idea would be to do something like the following (using the RangeChecker as an example). As I haven't used proc macros extensively, there are definitely issues with naming and syntax, but something along these lines should be doable

`
// CONSTANTS
// ================================================================================================
// Constants for the columns & constraint degrees - we have to define these in each processor's constraint module
...

// RANGE CHECKER CONSTRAINTS
// ================================================================================================
#[derive(Constraints)]
pub struct RangeCheckerConstraints {
    #[constraint_boundaries = [NONE, NONE, NONE, 0]]
    first_step: Vec<ColumnBoundary>,
    #[constraint_boundaries = [NONE, NONE, NONE, 65535]]
    last_step: Vec<ColumnBoundary>,
    #[constraint_degrees = CONSTRAINT_DEGREES]
    transitions: TransitionConstraints,
}

// initialization is derived
// basic accessors needed by ProcessorConstraints trait are derived

impl ProcessorConstraints for RangeCheckerConstraints {
    fn enforce_constraints<E: FieldElement>(&self, frame: &EvaluationFrame<E>, result: &mut [E]) {
        // Constrain the selector flags.
        let mut index = enforce_flags(frame, result);

        // Constrain the row transitions in the 8-bit section of the table.
        index += enforce_8bit(frame, &mut result[index..]);

        // Constrain the transition from 8-bit to 16-bit section of the table.
        enforce_16bit(frame, &mut result[index..]);
    }
}


// TRANSITION CONSTRAINT HELPERS
// ================================================================================================
// Helper functions for enforce_constraints - we have to define these in each processor's constraint module
...

// RANGE CHECKER FRAME EXTENSION TRAIT
// ================================================================================================

// We could derive the column accessor methods, but I'd need to play around with it a bit to figure out how.
// We'll need to define the intermediate variables & helper methods in each processor's constraint module
...
`

**Comment by bobbinth** [comment]
The sketch looks great! Makes me wish we already had something like that :)

One comment: for boundary constraints we can probably just use the `Assertion` struct since index of the last step can be passed in via constructor parameter, and we'll have all information needed to create assertions at construction time.

**Comment by grjte** [comment]
If the plan is to consider using macros in the future as described above but reduce boilerplate for now, then I actually don't think the constraint module structs are worth keeping either, so I've removed them as well to simplify.

**Review by bobbinth — COMMENTED** [review]
This is another partial review as I haven't reviewed the memory constraints themselves. The overall structure looks good (I left just one minor comment), and once we have the new macro-based structure, it should be fairly easy to make the transition.

I believe outstanding things before we can merge this PR are:
- Fix the current issue with transition constraints not being satisfied.
- Update to the next version of Winterfell (in a different PR) so that we can bring blowup factor back to 8 and fix unstable constraint degrees with random value injection at the end of the trace.
- For me to actually review the memory constraints.

Once the above is done, we can probably start working on the macro-based approach.

**Comment by grjte** [comment]
> I believe outstanding things before we can merge this PR are:
> Fix the current issue with transition constraints not being satisfied.

This is fixed.
>  Update to the next version of Winterfell (in a different PR) so that we can bring blowup factor back to 8 and fix unstable constraint degrees with random value injection at the end of the trace.

This is updated.
>  For me to actually review the memory constraints.

Everything should now be ready for review again.

**Review by bobbinth — APPROVED** [review]
Looks great! Thank you!

A comment for the future: there are still two things which bother me in a way we define constraints:
1. The readability is still far from ideal.
2. We probably have quite a few duplicated computations in there (e.g., when computing flags).

Macros may help with both of these, but I wonder how far we can get with them. It is almost like it would be great to have some kind of a "pre-processing" step which would take pure constraint descriptions, optimize them to extract common sub-expressions, determine the resulting constraint degrees, and output Rust code which is optimal.

Of course, if we have such a tool, we'd need to verify that it does the translations/optimizations correctly, and it also reminds me of the work I did on AirScript a while back. But maybe there is something simpler we can do which would get us most of the way there.

**Comment by bobbinth** [comment]
One more note on performance: adding `#[inline(always)]` attribute to all accessor methods of frame extensions shaves off almost 10% off constraint evaluation time.
